ships/0026-image-abstraction-layer.md

```diff
@@ -0,0 +1,232 @@
+| hashReferences    | A list of all imported versions of the Image. Sorted by import time descending (the first item in this list is the last imported version) |
+   
+The property `.status.hashReferences` is a list of imported image versions, The operator would
+hold up to twenty five imports per image. Each item in the list is composed of the following
+properties:
+
+| Name           | Description  |
+| -------------- | ------------ |
+| from           | Keeps a reference from where the image got imported |
+| importedAt     | Date and time of the import |
+| imageReference | Where this reference points to (by hash), may point to the mirror registry in case of a mirrored Image |
+
+### ImageImport status
+
+On a ImageImport status, one would be able to find information about the import attempts executed
+by the operator, it looks like:
+
+| Name           | Description  |
+| -------------- | ------------ |
+| hashReference  | Once the import has succeeded this is populated with the data about the image. Properties for hashReference are equal to Image.Status.HashReferences above. This is what is copied into the Image.Status.HashReferences list. |
+| importedAt     | A list of failures or successes during the ImageImport processing |
+
+The properties for the object `.status.hashReference` are exactly the same as for the objects
+inside `Image.Status.HashReferences`. As for importAttempts objects they contain the following
+properties:
+
+| Name    | Description                                                                          |
+| ------- | ------------------------------------------------------------------------------------ |
+| when    | Date and time of the import attempt                                                  |
+| succeed | A boolean indicating if the import was successful or not                             |
+| reason  | In case of failure (succeed = false), what was the error                             |
+
+The operator must try to import multiple times the same ImageImport in case of failures. These
+imports should implement exponential backoff to avoid “too many requests” from remote registries.
+After ten failures the operator should stop trying to process the ImageImport.
+
+
+### User Stories
+
+As an operator I want to be able to create a reference by hash to an image hosted in a remote
+registry so I can easily refer to it by its hash instead of by tag.
+
+As an operator I want to be able to mirror an image in my registry so I can take ownership of it
+and possibly gain some performance while pulling.
+
+
+### Implementation Notes
+
+To import images hosted in registries protected by authentication the operator would need to be
+able to read docker credentials secrets that exist in the same namespace where the Image exists.
+The idea is to mimic OpenShift’s Image Stream here: read all docker config secrets from the target
+namespace and use the appropriate one (if present).
```

Comment:
I see, related to my point above. I am not sure if this implicity is a good idea. One can easily have multiple secrets for the same host that have access to different repositories under that container registry host. Though, not sure if that is realistic inside a single Kubernetes namespace.

Reply:
Not sure where your concern is, if I understand you right then we are covered by a [keyring](https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/secrets/secrets.go#L29) as used by k8s itself, no? Can you think in a scenario where such a keyring would fall short?